File: problems/can-scramble/can-scramble.py
from collections import Counter

def can_scramble(source, dest):
  if len(source) != len(dest):
    return False
  return Counter(source) == Counter(dest)


assert(can_scramble("abc", "cba") == True)
assert(can_scramble("abc", "ccc") == False)
assert(can_scramble("aab", "bbc") == False)
assert(can_scramble("aabaaaa", "bbc") == False)

File: problems/can-scramble/can_scramble.rs
use std::collections::HashMap;

trait CharMap {
    fn char_map(self) -> HashMap<char, u32>;
}

impl<'a> CharMap for &'a str {
    fn char_map(self) -> HashMap<char, u32> {
        let mut map = HashMap::new();
        for c in self.chars() {
            map.entry(c).and_modify(|count| *count += 1).or_insert(1);
        }
        map
    }
}

pub fn can_scramble(input: &str, output: &str) -> bool {
    input.char_map() == output.char_map()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(can_scramble("", ""), true);
        assert_eq!(can_scramble("abc", "cba"), true);
        assert_eq!(can_scramble("aaa", "aa"), false);
        assert_eq!(can_scramble("aab", "bba"), false);
        assert_eq!(can_scramble("ac", "bb"), false);
    }
}
